src/bootstrap/builder.rs

```diff
@@ -1551,7 +1553,7 @@ impl<'a> Builder<'a> {
             Mode::ToolStd => {
                 // Right now this is just compiletest and a few other tools that build on stable.
                 // Allow them to use `feature(test)`, but nothing else.
-                rustflags.arg("-Zallow-features=binary-dep-depinfo,test,backtrace");
+                rustflags.arg("-Zallow-features=binary-dep-depinfo,test,backtrace,proc_macro_internals,proc_macro_diagnostic,proc_macro_span");
```

Comment:
This looks pretty suspicious to me -- the comment is here for a reason, we shouldn't be expanding this list to arbitrary features, since that means that tools will eventually need cfg(bootstrap) internally, which is painful to support (for us and for the tools). Is there a reason we're not just using Std for rust-analyzer? That seems much more appropriate; it clearly doesn't actually build on stable.

Reply:
Hmm
Isn't there lots of stuff in bootstrap that already switches on Mode::Std? I worry changing it will break something else.

I do agree the proc macro server is inherently unstable; I guess the best thing to do is to switch and see what breaks. Don't have time for that today.

Reply:
Ok, having thought about this more I think ToolStd is correct. If you look at when it was originally added, the motivation was to allow distros to build compiletest with the previous beta compiler. Before ToolStd, it would break on change to libtest.

So this does exactly what we want: it allows using nightly features from the standard library, but not language features in the nightly compiler. The fact that it happens to use the beta standard library for build scripts is just because the rustc shim is not very smart and does that unconditionally for all build scripts in stage 0.

Reply:
proc_macro_span and proc_macro_diagnostic are definitely language-y features, not just library features (even if their surface area is defined there). So I'm not sure I buy that argument fully.

In any case, it seems maybe OK to leave this until it causes us trouble down the line -- as long as rust-analyzer is a subtree it's not too hard to fix breakage caused by these features changing, hopefully.

Reply:
Sounds good. If it makes you feel better, this allows strictly fewer nightly features than before (it used to be ToolRustc) and the RA devs care a lot about being able to build on stable, so I think this is unlikely to break :)

Reply:
Well, clearly these features are being used by something that's not going to build on stable, and I'm much more likely to see breakage to them coming from modifications in rustc/std than from work on rust-analyzer, so this is not particularly reassuring :)

ToolRustc is in some ways *better*; it allows all nightly features, but we learn of breakage *immediately*. With ToolStd, we learn of breakage when we do a beta-bump, which is a terrible time. In rust-analyzer's case maybe we're lucky and they're already doing this same thing on their CI and will do a bump to rust-lang/rust in time, but there's a time and someone-is-running-CI-often component there that makes me concerned.

Reply:
> ToolRustc is in some ways _better_; it allows all nightly features, but we learn of breakage _immediately_

Huh, that might explain why `x.py test --stage 0 rust-analyzer` used to work in my first subtree PR. And it might be the better choice here eventually. 🤔 

Again though, until you think this is a blocker for the merge, for now I'm worried about merge conflicts that would cause me to redo the whole PR (I don't mind doing it if we have a compelling reason, but it is time I'd rather spend elsewhere).

Reply:
> Huh, that might explain why x.py test --stage 0 rust-analyzer used to work in my first subtree PR. 

That sounds unlikely. The shim uses beta libstd for all stage 0 build scripts regardless of the mode.

I am slightly unsure why @Mark-Simulacrum thinks we'd only catch breakage in beta, since this is only using features from std.

Reply:
Those features are thin shims over significantly larger surface area in the compiler (AFAIK), like the rest of proc-macro; the compiler used during the use of these features is presumably the stage0 beta, so you're not exercising any new code in the compiler until we bump that beta.

I continue to not believe this is a blocker for merging (otherwise I'd have r-'d), just flagging a concern, partially so that we can get on the same page and make sure @jyn514 is aware of this for future bootstrap changes, potentially unrelated to rust-analyzer. (Or that I'm aware that this is the wrong conclusion to draw, alternatively :)